### ISSUE TITLE
Extending Documentation: Saving custom models

### ISSUE THREAD

**Issue by HansBambel** [mlflow/mlflow#4230]
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for information on what types of issues we address.

**Please fill in this documentation issue template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages documentation fix contributions. Would you or another member of your organization be willing to contribute a fix for this documentation issue to the MLflow code base?

- [ ] Yes. I can contribute a documentation fix independently.
- [x] Yes. I would be willing to contribute a document fix with guidance from the MLflow community.
- [ ] No. I cannot contribute a documentation fix at this time.

### URL(s) with the issue:
[Model-customization](https://mlflow.org/docs/latest/models.html#model-customization)
[PyFunc-create-custom](https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#pyfunc-create-custom)

Please provide a link to the documentation entry in question.

### Description of proposal (what needs changing):
I am currently trying to log some of our custom models which are comprised of LightGBM models deep down but have other added functionalities as well.
By now, I am able to log the metrics and parameters manually, but I am having issues with logging artifacts and the models themselves.

I tried inheriting our models from [PythonModel](https://mlflow.org/docs/latest/models.html#model-customization), but this has three issues:
1. The abstract function `predict` from PythonModel also contains `context` and most of our models already implemented a `predict` function that does not require this additional parameter. 
2. One of our models requires additional kwargs, but with the current implementation of `PythonModel` this is not possible as far as I understand it (#2360).
3. I can not use the inherited `PythonModel` for inference before serializing (example below) which is very odd when I just want to extend my model to be able to be logged by MLflow.
```
import mlflow.pyfunc

# Define the model class
class AddN(mlflow.pyfunc.PythonModel):

    def __init__(self, n):
        self.n = n

    def predict(self, context, model_input):
        return model_input.apply(lambda column: column + self.n)

# Construct and save the model
model_path = "add_n_model"
add5_model = AddN(n=5)
mlflow.pyfunc.save_model(path=model_path, python_model=add5_model)

# Load the model in `python_function` format
loaded_model = mlflow.pyfunc.load_model(model_path)

# Evaluate the model
import pandas as pd
model_input = pd.DataFrame([range(10)])
# This works
model_output = loaded_model.predict(model_input)
# This does not work as it's missing the context parameter
model_output = add5_model(model_input)

assert model_output.equals(pd.DataFrame([range(5, 15)]))
```

To summarize, I would extend the documentation by answering the following questions:
1. Am I wrong in assuming that I can just convert my custom model to a `PythonModel` and expect it to work?
2. Is the `PythonModel`-class to be used after my custom-model is trained and I just want to log it? 
3. What is the connection between `PyFuncModel` and `PythonModel`?
4. What is the `context` and why do I need it in `PythonModel` and not in `PyFuncModel`?

**Comment by dmatrix** [comment]
@HansBambel Thanks for raising this issue. It's on my list of adding more docs with respect to Python func flavor. I just merged a[ PR](https://github.com/mlflow/mlflow/pull/4225)

Next on my list is to explain at length the use case of the PyFunc Model, provide some examples in the dir examples/pyfunc/, and provide an explanation for `context`, and the difference between PythonModel and PyFuncModel. This includes saving customer models wrapped in Pyfunc flavor and able to log and register them.

But we welcome contributions!
